bus-mapping/src/evm/opcodes/begin_end_tx.rs

```diff
@@ -313,17 +314,62 @@ fn gen_end_tx_steps(state: &mut CircuitInputStateRef) -> Result<ExecStep, Error>
         false,
     )?;
 
+    end_tx(state, &mut exec_step, &call)?;
+
+    Ok(exec_step)
+}
+
+pub(crate) fn begin_tx(
+    state: &mut CircuitInputStateRef,
+    exec_step: &mut ExecStep,
+    call: &Call,
+) -> Result<(), Error> {
+    // Write the transaction id
+    state.call_context_write(
+        exec_step,
+        call.call_id,
+        CallContextField::TxId,
+        state.tx_ctx.id().into(),
+    )?;
+    Ok(())
+}
+
+pub(crate) fn end_tx(
+    state: &mut CircuitInputStateRef,
+    exec_step: &mut ExecStep,
+    call: &Call,
+) -> Result<(), Error> {
+    // Write the tx receipt
+    write_tx_receipt(state, exec_step, call.is_persistent)?;
```

Comment:
Should we write a tx receipt if it's invalid?

Reply:
I write the receipt because `end_tx` expects them to be there for the previous transaction:

https://github.com/privacy-scaling-explorations/zkevm-circuits/blob/7617394711a0e79beec5ef1f792938242be3dbcb/zkevm-circuits/src/evm_circuit/execution/end_tx.rs#L304-L310

So if they are not there in the same format like with valid transactions, I think it would create a messy situation between valid and invalid transactions.

It's used to track the `CumulativeGasUsed` over multiple transactions, so we still have to do that at least for invalid transactions as well (though no update to the value because no gas is used).